cf-agent/verify_processes.c

```diff
@@ -231,12 +233,29 @@ static PromiseResult VerifyProcessOp(EvalContext *ctx, const Attributes *a, cons
             {
                 if (IsExecutable(CommandArg0(a->process_stop)))
                 {
-                    ShellCommandReturnsZero(a->process_stop, SHELL_TYPE_NONE);
+                    Log(LOG_LEVEL_DEBUG, "Found process_stop command '%s' is executable.", a->process_stop);
+
+                  if (ShellCommandReturnsZero(a->process_stop, SHELL_TYPE_NONE))
+                  {
+                      cfPS(ctx, LOG_LEVEL_INFO, PROMISE_RESULT_CHANGE, pp, a,
+                           "Promise to stop '%s' repaired, '%s' returned zero",
+                           pp->promiser, a->process_stop);
+                      result = PromiseResultUpdate(result, PROMISE_RESULT_CHANGE);
+                      return result;
```

Comment:
You're introducing a memory leak here, by not calling `DeleteItemList()` before returning. Look at the other return statements in the same functions to see what cleanup is necessary.

Also, AFAICT you shouldn't return here, there is more logic later, that would run after the shell command. Most notably related to signals and restarting.

Reply:
Yeah, I returned from here because of where the else was already returning. I see that there it had DeleteItemList(killlist) though, so Ill add that in.

Reply:
Seems like before  doing the signals (i guess that's what's happening here)

https://github.com/cfengine/core/pull/3855/files#diff-6c1d2f248881f255c48f13ff8f1d5fc6L248

it should only try to signal the processes that still exist, like the process table should be refreshed. It's my current understanding that the process table is only ever refreshed at the start of a bundle. But it seems like if we knowingly alter a process, we should refresh our view of the process state as we do with our package cache.

Reply:
I admit, this code is very hard to completely understand, but I think it's always safer to preserve behavior. Clearly, there are not enough tests for this code, but it might be working for people in real policy, and we don't want to break that.

So, I would remove this early return (and the deleteitemlist), at least here, since this is the success path.
Previously it would **not** return early in any of these cases you added. There is more code below for signalling and restarting. With your changes it **always** returns early after the shell command.

Reply:
I found in the changelog that 3.3.0 introduced the ability for process\_stop and
signals to co-exist within the same promise.

> -   Allow both \`process\_stop' and \`signals' constraints in
>     \`processes' promises at the same time.

This makes me think that it should *not* return early. It should do both
process\_stop and signals regardless of success or failure in either.

But I guess that it has not been sending signals if process\_stop was not
executable, so I have only removed the early returns that I added.